StrongMind/canvas-lms
# Qty 307

[QTY-307](https://strongmind.atlassian.net/browse/QTY-307)

## Purpose

To increase the usability of Conferences, add the ability to filter users by role when creating a new Conference or editing an existing one. 

## Approach

Add checkboxes that allow teachers to filter the User list when creating/editing a Conference by their role in the course and add them to the Conference (screenshots below). 

Due to the complexity of the JavaScript code necessary to make this feature work with AJAX/CoffeeScript/Handlebars/Backbone.js, we ended up simply adding code to hide and reveal the HTML elements for selection. This ended up being a much simpler solution to implement than the original plan of using AJAX to repopulate the list and reduced the number of queries necessary to accomplish the task. 

In order to make the filters function as expected, we needed to change the verbiage and default behavior of Instructures' `select all` checkbox. The User list is now always displayed rather than hidden by default with the `invite all` option selected. 

Additionally, we needed to update the `ConferencesController#web_index` action to query `Enrollments` so that a `User` could have their `Role` mapped to the JSON used to populate the list for filtering. We also added a query for `Roles` so that we could create the list of possible filters and mapped that to the `js_env`. 

At this time, it was indicated that we should only include filters for `Teacher`, `Student`, `Observer`, and `TA` roles.


## Testing

1. Create a course in enterprise Canvas and add `Users` with various `Roles` to the course. (It is helpful to omit at least one of the possible roles so that the case where no `User` matches the selected filter can be tested.)
2. Create a new `Conference` and use the Role filters to confirm that only the expected users appear in the list.
3. Edit the conference and confirm that the previously added users have a `disabled` checkbox next to their name which remains disabled even after filtering. 
4. Update the conference with a new user. Refresh the page and confirm that they have been added to the existing conference and now have a disabled checkbox next to their name in the list. 

## Screenshots

### List displayed by default with checkbox filters: 

<img width="1725" alt="image" src="https://user-images.githubusercontent.com/30609917/196737375-3a512d8a-113e-4ecb-8ccb-34b12c7cb074.png">


### Filtered with one Role selected:

<img width="1727" alt="image" src="https://user-images.githubusercontent.com/30609917/196737637-8e6ec68a-9aaa-4b20-971f-820a3ac72b3a.png">

### Filtered with multiple Roles selected:

<img width="1728" alt="image" src="https://user-images.githubusercontent.com/30609917/196737727-eb922267-6907-4e93-92ec-0177dc3d6e6c.png">

### Filtered with a Role that matches no users in the course displays error messaging: 

<img width="1728" alt="image" src="https://user-images.githubusercontent.com/30609917/196738000-be581ce5-0a8f-4be7-9183-36bffe77a4cf.png">

### Users are still disabled when editing a Conference:

<img width="1728" alt="image" src="https://user-images.githubusercontent.com/30609917/196738378-e037c68e-5c0a-450d-b9e3-92a04a0bd8d3.png">
